File: java_hummus/src/test/java/com/github/klee0kai/hummus/collections/weaklist/ChangeListsHelper.java
package com.github.klee0kai.hummus.collections.weaklist;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import com.github.klee0kai.hummus.interfaces.ICallBackResult;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;

public class ChangeListsHelper {

    public static <T, R> void changeForEachListsSame(List<List<T>> lists, ICallBackResult<List<T>, R> callable) {
        R qualityResult = null;
        for (List<T> l : lists) {
            R res = callable.call(l);

            if (qualityResult != null) assertEquals(qualityResult, res);
            else qualityResult = res;
        }
    }


    public static <T> void assertListsSame(List<T> expected, List<T> actual, T[] emptyArray) {
        assertEquals(expected.isEmpty(), actual.isEmpty());
        assertEquals(expected.size(), actual.size());

        assertEquals(
                Arrays.asList(expected.toArray()),
                Arrays.asList(actual.toArray())
        );
        assertEquals(
                Arrays.asList(expected.toArray(emptyArray)),
                Arrays.asList(actual.toArray(emptyArray))
        );


        {
            // iterators are same
            Iterator<T> expIt = expected.iterator();
            Iterator<T> actIt = expected.iterator();
            while (expIt.hasNext()) {
                assertTrue(actIt.hasNext());

                T extItem = expIt.next();
                T actItem = actIt.next();
                assertEquals(extItem, actItem);
                assertEquals(expected.indexOf(extItem), actual.indexOf(actItem));
                assertEquals(expected.lastIndexOf(extItem), actual.lastIndexOf(actItem));
                assertEquals(expected.contains(extItem), actual.contains(actItem));
            }
            assertFalse(actIt.hasNext());
        }

        {
            // listiterators are same
            Iterator<T> expIt = expected.listIterator();
            Iterator<T> actIt = expected.listIterator();
            while (expIt.hasNext()) {
                assertTrue(actIt.hasNext());

                T extItem = expIt.next();
                T actItem = actIt.next();
                assertEquals(extItem, actItem);
                assertEquals(expected.indexOf(extItem), actual.indexOf(actItem));
                assertEquals(expected.contains(extItem), actual.contains(actItem));
            }
            assertFalse(actIt.hasNext());
        }

        //contains each other
        assertTrue(actual.containsAll(expected));
        assertTrue(expected.containsAll(actual));
    }

    public static class SimpleIntPair {
        int a, b;

        public SimpleIntPair(int a) {
            this.a = a;
        }

        public SimpleIntPair(int a, int b) {
            this.a = a;
            this.b = b;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            SimpleIntPair that = (SimpleIntPair) o;
            return a == that.a && b == that.b;
        }

        @Override
        public int hashCode() {
            return Objects.hash(a, b);
        }
    }

}

File: kdevkit/src/test/kotlin/com/github/klee0kai/android_devkit/collections/weaklist/ChangeListsHelper.kt
package com.github.klee0kai.android_devkit.collections.weaklist

import org.junit.Assert.*

object ChangeListsHelper {

    fun <T, R> changeForEachListsSame(
        vararg lists: MutableList<T>,
        block: MutableList<T>.() -> R
    ) {
        var qualityResult: R? = null
        for (l in lists) {
            val res: R = block(l)
            if (qualityResult != null) assertEquals(qualityResult, res) else qualityResult =
                res
        }
    }

    inline fun <reified T> assertListsSame(
        expected: List<T>,
        actual: List<T>,
    ) {
        assertEquals(expected.isEmpty(), actual.isEmpty())
        assertEquals(expected.size.toLong(), actual.size.toLong())
        assertEquals(
            expected.toTypedArray().toList(),
            actual.toTypedArray().toList(),
        )
        run {

            // iterators are same
            val expIt = expected.iterator()
            val actIt = expected.iterator()
            while (expIt.hasNext()) {
                assertTrue(actIt.hasNext())
                val extItem = expIt.next()
                val actItem = actIt.next()
                assertEquals(extItem, actItem)
                assertEquals(
                    expected.indexOf(extItem).toLong(),
                    actual.indexOf(actItem).toLong()
                )
                assertEquals(
                    expected.lastIndexOf(extItem).toLong(),
                    actual.lastIndexOf(actItem).toLong()
                )
                assertEquals(expected.contains(extItem), actual.contains(actItem))
            }
            assertFalse(actIt.hasNext())
        }

        run {
            // listiterators are same
            val expIt: Iterator<T> = expected.listIterator()
            val actIt: Iterator<T> = expected.listIterator()
            while (expIt.hasNext()) {
                assertTrue(actIt.hasNext())
                val extItem = expIt.next()
                val actItem = actIt.next()
                assertEquals(extItem, actItem)
                assertEquals(
                    expected.indexOf(extItem).toLong(),
                    actual.indexOf(actItem).toLong()
                )
                assertEquals(expected.contains(extItem), actual.contains(actItem))
            }
            assertFalse(actIt.hasNext())
        }

        //contains each other
        assertTrue(actual.containsAll(expected))
        assertTrue(expected.containsAll(actual))
    }

}